platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/Tipset.java

```diff
@@ -0,0 +1,199 @@
+import com.swirlds.common.system.address.AddressBook;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Represents a slice of the hashgraph, containing one "tip" from each event creator.
+ */
+public class Tipset {
+
+    private final AddressBook addressBook;
+
+    /**
+     * The tip generations, indexed by node index.
+     */
+    private final long[] tips;
+
+    /**
+     * Create an empty tipset.
+     *
+     * @param addressBook the current address book
+     */
+    public Tipset(@NonNull final AddressBook addressBook) {
+        this.addressBook = Objects.requireNonNull(addressBook);
+        tips = new long[addressBook.getSize()];
+
+        // Necessary because we currently start at generation 0, not generation 1.
+        Arrays.fill(tips, -1);
+    }
+
+    /**
+     * Build an empty tipset (i.e. where all generations are -1) using another tipset as a template.
+     *
+     * @param tipset the tipset to use as a template
+     * @return a new empty tipset
+     */
+    private static @NonNull Tipset buildEmptyTipset(@NonNull final Tipset tipset) {
+        return new Tipset(tipset.addressBook);
+    }
+
+    /**
+     * <p>
+     * Merge a list of tipsets together.
+     *
+     * <p>
+     * The generation for each node ID will be equal to the maximum generation found for that node ID from all source
+     * tipsets.
+     *
+     * @param tipsets the tipsets to merge, must be non-empty, tipsets must be constructed from the same address book or
+     *                else this method has undefined behavior
+     * @return a new tipset
+     */
+    public static @NonNull Tipset merge(@NonNull final List<Tipset> tipsets) {
+        Objects.requireNonNull(tipsets, "tipsets must not be null");
+        if (tipsets.isEmpty()) {
+            throw new IllegalArgumentException("Cannot merge an empty list of tipsets");
+        }
+
+        final int length = tipsets.get(0).tips.length;
+        final Tipset newTipset = buildEmptyTipset(tipsets.get(0));
+
+        for (int index = 0; index < length; index++) {
+            long max = -1;
+            for (final Tipset tipSet : tipsets) {
+                max = Math.max(max, tipSet.tips[index]);
+            }
+            newTipset.tips[index] = max;
+        }
+
+        return newTipset;
+    }
+
+    /**
+     * Get the tip generation for a given node ID.
+     *
+     * @param nodeId the node ID in question
+     * @return the tip generation for the node ID
+     */
+    public long getTipGenerationForNodeId(@NonNull final NodeId nodeId) {
+        final int index = addressBook.getIndexOfNodeId(nodeId);
+        return tips[index];
+    }
+
+    /**
+     * Get the tip generation for a given node
+     *
+     * @param nodeId the node in question
+     * @return the tip generation for the node
+     */
+    public long getTipGenerationForNode(@NonNull final NodeId nodeId) {
```

Comment:
This method is a duplicate of the previous

Reply:
Lol, so it is. Deduped.